### ISSUE TITLE
RStudio hangs up while working on visual markdown

### ISSUE THREAD

**Issue by llrs** [rstudio/rstudio#10823]
### System details

    RStudio Edition :  Desktop
    RStudio Version :  2022.02
    OS Version      : Ubuntu 20.04
    R Version       : 3.6.3

### Steps to reproduce the problem

I haven't found a way to reproduce the problem yet.

### Describe the problem in detail

On visual markdown sometimes the IDE freezes, with a banner, with Working... for more than 15 minutes I've waited, but had to close and continued working after restarting it. Other times it happens when I move the cursor over a citation. 
To restart it doesn't work to close from the window, I need to do it from outside RStudio.

While the Working... banner appears I cannot edit the document, open any new window from the menus.
When I quit, the modifications are lost (not saved even if I click save on the menu it appears). 
I thought this problem might be related to the .Rproj.user folder, but right before opening it I had removed it completely. 

It might be related to https://github.com/rstudio/rstudio/issues/9878 but I think it might be different so I opened this issue ( https://github.com/rstudio/rstudio/issues/9878#issuecomment-1075047713) . 

Last time there was this message on the terminal (I opened the project from the terminal):

```
"WARNING: 3 chunks in scope tree, but 2 chunks in visual editor.\n\tat Unknown.new Vk(rstudio-0.js)\n\tat Unknown.A_d(rstudio-0.js)\n\tat Unknown.Eem(rstudio-0.js)\n\tat Unknown.lfm(rstudio-0.js)\n\tat Unknown.Hhm(rstudio-0.js)\n\tat Unknown.S$l(rstudio-0.js)\n\tat Unknown.IE(rstudio-0.js)\n\tat Unknown.yE(rstudio-0.js)\n\tat Unknown.KXk(rstudio-0.js)\n\tat Unknown.j0k(rstudio-0.js)\n\tat Unknown.tc(rstudio-0.js)\n\tat Unknown.eval(rstudio-0.js)\n\tat Unknown.Cm(rstudio-0.js)\n\tat Unknown.Fm(rstudio-0.js)\n\tat Unknown.eval(rstudio-0.js)\n"
"The Web Audio autoplay policy will be re-enabled in Chrome 70 (October 2018). Please check that your website is compatible with it. https://goo.gl/7K7WLu"
```

The logging file `~/.local/share/rstudio/log/rdesktop.log` had this from a previous hang up:
```
2022-01-28T11:21:38.166594Z [rdesktop] ERROR Main window closed unexpectedly; LOGGED FROM: rstudio::desktop::MainWindow::closeEvent(QCloseEvent*)::<lambda(QVariant)> src/cpp/desktop/DesktopMainWindow.cpp:524
```

Not sure the audio problem comes from, but I was hearing some music on the default browser (Chromium).
Chromium: Version 99.0.4844.82 (Official Build) snap (64-bit)
Chrome:     Version 99.0.4844.82 (Official Build) (64-bit)

<details>
<summary>The project I was working on</summary>

https://github.com/llrs/thesis

</details>

### Describe the behavior you expected

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [ ] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@llrs Thank you for raising! Sorry this is causing such a problem, and of course you shouldn't have to start from the terminal. I'm marking this as an issue to review as continue work on RStudio, especially the visual markdown editor.

**Comment by llrs** [comment]
Today I opened the same project from the terminal.
I had been working for several hours without problems. But after adding a new reference and modifying it to prevent duplicating the reference (Rstudio reference manager inserted [@citationKeya] instead of [@citationKey] [Maybe I selected incorrectly the entry]) it hang up with a "Working..." banner.
![Screenshot from 2022-04-01 of the screen of Rstudio with the banner and the citation I edited at the bottom](https://user-images.githubusercontent.com/6818218/161274788-dc4874a6-a6f5-43c3-9cdb-206d191ff06d.png)
 
I've waited for 15 minutes and it hasn't solved itself. I can scroll or resize the different sections but no edit a file or change between opened files.

**Comment by llrs** [comment]
I was working on the same project with a different computer (Ubuntu 20.04 and R 4.1.2) and it froze in a similar way. Reporting this because I was starting to think it was related to some configuration on my other computer.

**Comment by KevinShook** [comment]
Hi, I just want to confirm that I have been having exactly the same issue as llrs.
I get the same error messages:
```

**Comment by llrs** [comment]
Thanks for reporting back. ~~Much~~ All of my writing is already done so I probably won't run long enough to trigger this again (if still exists), but I will report back if it happens again.

**Comment by mikebessuille** [comment]
@llrs if you encounter this again, please comment.  I'll close this now since we believe it's been addressed in the latest (Spotted Wakerobin).

**Comment by lucblassel** [comment]
HI @mikebessuille, I have been encountering this issue as well, I am using `RStudio 2022.02.0+443 "Prairie Trillium" ` on MacOS. 

When working on Rmarkdown files in visual mode for a bookdown project it often gets hung up and the small white banner at the top shows either "working" or "loading collections" and I have to quit RStudio, losing unsaved changes. 

It seems to happen more often when working with citations through the zotero integration but I have not been able to reliably reproduce this behaviour. 

Here is my system info:
```
R version 4.2.0 (2022-04-22)
Platform: x86_64-apple-darwin21.3.0 (64-bit)
Running under: macOS Monterey 12.4
```

ANd full RStudio info:
```
RStudio 2022.02.0+443 "Prairie Trillium" Release (9f7969398b90468440a501cf065295d9050bb776, 2022-02-16) for macOS
Mozilla/5.0 (Macintosh; Intel Mac OS X 12_4_0) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.10 Chrome/69.0.3497.128 Safari/537.36
```

**Comment by llrs** [comment]
Sorry to hear, that this still affects you @lucblassel . About the citations, I didn't mention but I was also using Zotero to access my bibliography.

**Comment by Entomophagist** [comment]
I'm having a very similar problem when working with citations from Zotero. I had been suspicious that the problem was arising from the add-in RBBT, but even when I removed the package, the problem persisted.

RStudio info:
```
"Prairie Trillium" Release (1db809b8, 2022-05-20) for Ubuntu Bionic
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.8 Chrome/69.0.3497.128 Safari/537.36
```
R info:
```
R version 4.2.0 (2022-04-22) -- "Vigorous Calisthenics"
Platform: x86_64-pc-linux-gnu (64-bit)
```

Zotero info:
`Version 6.0.9`

The problem occurs reliably for me when interacting with citations. For example, I just reproduced the problem with the following steps:

1. I open my thesis.Rmd file. No problem. I can edit and save the body text of the document without provoking the freeze.
2. I click on the text of any in-line citation. As expected, a tooltip immediately appears containing information about the reference. 
![image](https://user-images.githubusercontent.com/95774269/175698287-363b075d-c456-4664-b29f-423a45106910.png)

3. I change the text of a citation. I can input gibberish, or the tag of a real reference.
![image](https://user-images.githubusercontent.com/95774269/175698368-ab85b75e-6f98-4351-8303-53d29730e6da.png)

4. I click on the text of any in-line citation. No tooltip appears. After a pause (~5 seconds), I get an overlapping cascade of notifications at the toolbar of "Reading Collections" then "Loading Citations", repeating a few times and either getting stuck on "Loading Citations" or finally on "Working".
![image](https://user-images.githubusercontent.com/95774269/175697117-0f2d18d1-589f-46ed-9056-fe0d84a2f135.png).

At this point, I can still navigate the Rmd document and make changes, but I cannot save, I cannot use any of the tools on the toolbar, and I cannot open any new files. Interestingly, clicking on any citation causes a "Reading bibliography..." notification to pop up and then another "Loading Collections..." notification.

Any help with this would be appreciated!

**Comment by JohannPopp** [comment]
> There is an issue just like this which occurred as a result of some 3rd party packaged Linux versions of RStudio not yet having some fixes we had in a patch release (this for sure happened with the arch package, not sure about others).
> 
> This should be resolved with the very latest patch release of RStudio (2022.02.2+485) which was released last week so should be available in whatever archive you are getting RStudio from.

The issue seems to happen also with release RStudio 2022.07.2+576 "Spotted Wakerobin" Release (e7373ef, 2022-11-21) for x86_64 GNU/Linux on Manjaro via AUR: rstudio-desktop.

Kind regards
Johann

**Comment by ronblum** [comment]
@JohannPopp Is this with Zotero citations or any citations?

**Comment by JohannPopp** [comment]
Dear Ron,

thank you for asking. It actually happened as soon as I tried to switch to the visual mode. In the end it seems to be an issue of this special AUR package "rstudio-desktop". 

I had switched to that package, because my former installation of "rstudio-desktop-bin" did not start any more throwing the message `error while loading shared libraries: libcrypto.so.1.1: cannot open shared object file: No such file or directory`. 

After switching back to "rstudio-desktop-bin" and solving the described problem by installing `openssl-1-1` everything works fine.

Kind regards
Johann

**Comment by dineiar** [comment]
I am seeing this same issue in RStudio Server. The full environment information is:

> RStudio 2022.02.1+461 "Prairie Trillium" Release (8aaa5d470dd82d615130dbf663ace5c7992d48e3, 2022-03-17) for Ubuntu Bionic
> Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36

It seems to be triggered when working with citations: I have a simple `.bib` file with a few entries and it hangs every time I mouse over a citation or position the cursor right after a `@` character.

**Comment by mikebessuille** [comment]
It seems this may still be happening.  Some comments above note that it happens on Prairie Trillium (which is known); please try to reproduce on the 2nd patch of PT, 2022.02.2, or later.  @JohannPopp has reproduced this with Spotted Wakerobin (2022.07.2) so it's likely still an issue.

**Comment by mikebessuille** [comment]
Per JJ: this is a couple weeks' of work, related to Zotero web collections.   Can't do this for Cherry Blossom.

**Comment by mikebessuille** [comment]
JJ suggests you talk with Charles when fixing this; it may involve turning a blocking into a non-blocking call, but this might trigger other changes that Charles would be aware of.

**Comment by rich-rstudio** [comment]
From a support [ticket](https://rstudioide.zendesk.com/agent/tickets/87326) - Posit workbench stops responding when they try to use the Visual editor for editing RMD templates. Some users have no issue on certain Rmd files which are used for rendering reports (created in desktop R studio). But they have issues if the files are used for generating HTML documentation (or may be created in RSP). The problem seems to be happening when working on projects synced to github. When they move the files to some other local home directory and sync github folders without these Rmd templates things are working fine.

**Comment by jlhubdata** [comment]
Adding a related issue: #87326

**Comment by timtmok** [comment]
I have reproduced the issue with latest from `main`. It appears to occur on Linux. I could not reproduce on Mac. After inserting a citation and clicking on it, it started "reading bibliography.." and I can no longer save the file.

**Comment by timtmok** [comment]
I can't seem to reproduce it reliably and have only done so twice so far. I have noticed that moving the cursor around will make a call for each character traversed to retrieve the bibliography. This doesn't seem optimal but does eventually resolve.

The last call before hanging is to get Quarto cross references. I've got the latest visual editor for 2023.05.0 and it includes fixes where Quarto cross references were causing hangs with the visual editor. There might be another issue with getting cross references but I haven't determined the scenario that causes it.

No workaround is available at the moment.

**Comment by timtmok** [comment]
It turns out that I'm running into an arm64 specific issue since Quarto isn't enabled yet. The code to detect cross references returns unexpected data to the visual editor, which causes the hang. This is specifically a different issue and would only occur with arm64 builds.

**Comment by timtmok** [comment]
I have not been able to reproduce after getting Quarto set up on arm64. I suspect this may have been fixed recently by changes in #12882. Although, I have not been able to reproduce this in 2022.02.2.

If any user does come across this again and can reproduce this easily, it would be helpful to have the data from the network activity. It can be retrieved by:
1. Right click anywhere in the IDE
2. Choose "Inspect Element" to open the dev tools window
3. In the dev tools window, switch to the network tab
4. Reproduce the issue

Back in the dev tools window, it will have an activity log. A screenshot of this screen attached here would be helpful. It would look something like this:


![Image](https://github.com/rstudio/rstudio/assets/9591545/9c9d87d2-7320-42ad-8848-2026491114c7)

I'd be interested in the highlighted events. I suspect the call to `quarto_xref_for_id` has an issue that has been fixed in 2023.03.1 and 2023.05.0. Trying either of those versions would confirm if it has been fixed.

**Comment by llrs** [comment]
I didn't explicitly mention but this was originally found in Rmarkdown (bookdown) not quarto (In case this is a different bug and rmarkdown_xref_for_id or something similar should be looked up).

Thanks for looking into it.

**Comment by timtmok** [comment]
Yes, it sounds odd that Quarto cross references are retrieved during this. I haven't confirmed if this is necessary for Rmarkdown files or if it even is the cause of the hang. It is called in 2022.02.2 but would only be a problem for Linux arm64 users.

**Comment by llrs** [comment]
If it helps my computer is amd64 (with `dpkg --print-architecture`)

**Comment by cmmrandau** [comment]
I'm having a similar issue in RStudio with zotero integration on Fedora 38 (rstudio-desktop-2023.03.0+386-4.fc38.x86_64 ). It worked before so I'm not sure what's causing the issue. I can insert quotations with @... but when I hoover over an already existing quotation in visual mode it gets stuck with a banner "Loading collections".

**Comment by timtmok** [comment]
@cmmrandau Thank you for checking with 2023.03.1. I believe this has been fixed by #12882.

**Comment by cmmrandau** [comment]
Still the same problem in Fedora's rstudio-desktop-2023.06.0+421-1.fc38.x86_64 , works fine in the native version from rstudio's webpage.

**Comment by ronblum** [comment]
@cmmrandau This package is managed through the Fedora Project, so any issues specific to the Fedora pckage should be addressed with them.

**Comment by AAttllasss** [comment]
I'm encountering a similar issue. Previously, all my citations worked perfectly, but lately, whenever I try to add a new citation or reference an existing one, Visual editor freezes, displaying "Working" and "Reading bibliography." Interestingly, when I reopen the file and knit it, all the previously added citations work perfectly. Does anyone have any suggestions?

**Comment by eliocamp** [comment]
I can confirm that I'm having this issue with a new installation of R studio. The Visual editor stops responding with the "Reading bibliography" banner.

**Comment by ronblum** [comment]
@AAttllasss and @eliocamp what versions of RStudio are you using--version number and desktop/server?

**Comment by ronblum** [comment]
@AAttllasss and @eliocamp Also, do you know what's the last version that worked for you? Thanks!

**Comment by eliocamp** [comment]
I'm using the latest version of RStudio (2024.04.0 Build 735). I don't know which version I was using before, but probably the release just before this one. 

Here's a screencast of the issue: 

[Screencast from 10-05-24 12:09:58.webm](https://github.com/rstudio/rstudio/assets/8617595/99315848-3be7-467d-a1b0-42a6119dec03)

**Comment by wyuill** [comment]
I'm having the same error on Posit Workbench 2023.12.0+369.pro3, “Ocean Storm” (10c1b662) on Ubuntu 20.04.6 LTS. A couple of additional notes that may be of use:
- when the qmd file is not saved I have no issue adding references but when I try to save a new file with references or when I add references to an already saved file the error occurs and I'm forced to terminate the session to recover
- inserting references using visual model does work when using the toolbar (Insert>Citation or CTRL+SHIFT+F8) just not doing '@' in visual model.

**Comment by cmmrandau** [comment]
This problem still persists in 2024.04.1+748 (Ubuntu 24.04 LTS). Hovering or scrolling past citation causes the program to hang, saying "Reading bibliography" and "Working" in what seems to be a large number of banners being produced on top of each other.

**Comment by eliocamp** [comment]
Update that today this bit me when writing the NEWS.md file on a package because I used the @ symbol to tag a GitHub username.

**Comment by mbellitti** [comment]
Same problem happening to me on Fedora 40 (Framework laptop with intel CPU): while editing a `.qmd` document, if I type `@` RStudio shows the "Working..." banner and hangs.  

RStudio version: RStudio 2024.04.2+764 "Chocolate Cosmos" Release (e4392fc9ddc21961fd1d0efd47484b43f07a4177, 2024-06-05) for RHEL 9
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) rstudio/2024.04.2+764 Chrome/120.0.6099.291 Electron/28.3.1 Safari/537.36, Quarto 1.4.555

I have Zotero installed, with about ~400 entries in the library and Better BibTeX installed.

**Comment by kevinushey** [comment]
Would anyone be willing to share a Zotero library that I can try to use locally to reproduce these issues? (Or, does such an example library exist online?)

**Comment by mbellitti** [comment]
The problem persists even with an empty library. To test it I created a new profile (instructions [here](https://www.zotero.org/support/kb/multiple_profiles)). So to reproduce it's enough to

1. Install Zotero
2. In RStudio, create an empty `.qmd` document and type `@`. 

After step 2. I get the "Working..." banner that doesn't go away.

I also tested it adding a single entry to the library, and the behavior is the same.

**Comment by mbellitti** [comment]
I checked the logs (in RStudio under help > diagnostics > show log files) and there was an error message about missing `pandoc`:
```
2024-06-20T19:52:53.254733Z [rsession-myusername] ERROR system error 2 (No such file or directory) [exe: /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835
2024-06-20T19:52:53.254733Z [rsession-myusername] ERROR system error 2 (No such file or directory) [exe: /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835
```
so I installed it (at system level) and cross-referencing in quarto now works correctly for the library with a single entry.  

I switched back to the Zotero profile with my full library (~500 entries) and I kept getting the same error. Reinstalling RStudio seems to have fixed it. In summary:

1. Check the logs. If there's an error about missing pandoc,
2. Install pandoc,
3. Reinstall RStudio.

This feels like overkill and I don't understand how the error originally happened, but at least it fixed it in my case.

**Comment by eliocamp** [comment]
I'm checking and having the same problem

```
2024-06-21T02:08:11.999333Z [rsession-username] ERROR system error 2 (No such file or directory) [exe: /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835
```
However, I have pandoc installed.

**Comment by mbellitti** [comment]
RStudio version: RStudio 2024.04.2+764 "Chocolate Cosmos" Release (e4392fc
<https://github.com/rstudio/rstudio/commit/e4392fc9ddc21961fd1d0efd47484b43f07a4177>,
2024-06-05) for RHEL 9
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)
rstudio/2024.04.2+764 Chrome/120.0.6099.291 Electron/28.3.1 Safari/537.36,
Quarto 1.4.555

From the error message it's clear that RStudio is looking for a bundled
version of quarto and pandoc, I don't know why installing the pandoc
package on my system helped.

In any case if RStudio can't find the bundled pandoc shouldn't it raise an
error?

**Comment by kevinushey** [comment]
Unfortunately, I'm not able to reproduce even on my RHEL 9 VM (using the latest release of RStudio from https://posit.co/download/rstudio-desktop/).

In case it's useful:

<img width="855" alt="Screenshot 2024-06-21 at 1 18 19 PM" src="https://github.com/rstudio/rstudio/assets/1976582/a9f2a591-d816-4aac-a6eb-8338896f9c3a">

**Comment by ismaRP** [comment]
I had commented on a similar issue in this thread #14549
I tried reinstalling RStudio and Pandoc several times, but what worked was reinstalling Quarto manually to the latest version (1.5.54). Then Rstudio automatically detected it, since it's a newer one than the one that comes with it (as far as I know).

**Comment by bastooo** [comment]
In my log I can find the following line:

> 2024-08-15T14:02:45.022581Z [rsession-basti] ERROR system error 2 (Datei oder Verzeichnis nicht gefunden) [exe: /usr/lib/rstudio/bin/pandoc/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835

When I checked the directory pandoc is not in 
> /usr/lib/rstudio/bin/pandoc/
but in
> /usr/lib/rstudio/bin/pandoc/bin/tools/

After creating a soft link in the /usr/lib/rstudio/bin/pandoc/ it works like a charm:
`ln -s /usr/bin/pandoc /usr/lib/rstudio/bin/pandoc/pandoc`

I'm using Manjaro (arch based linux-distro), so maybe this info is helping out someone in a similar situation.

**Comment by kevinushey** [comment]
I think I was able to reproduce. From Help -> Diagnostics -> Request Log, it looks like `quarto_xref_for_id` is getting called, but that invocation isn't ever returning a response.

<img width="583" alt="Screenshot 2024-08-28 at 3 30 31 PM" src="https://github.com/user-attachments/assets/bdc6a6e6-22b0-45fb-8984-3cefc68a8c61">

**Comment by kevinushey** [comment]
Okay, I think I can reproduce. It's as simple as creating a citation like `@abc` in an R Markdown document in visual mode, and then clicking on that citation. It looks like we try to query Quarto for some information about the citation link, but that query is failing and stalling / crashing the R session?

**Comment by kevinushey** [comment]
Relevant part of GDB backtracke:

```
#0  futex_wait (private=<optimized out>, expected=14, futex_word=0xaafc60de87dc) at ../sysdeps/nptl/futex-internal.h:146
#1  futex_wait_simple (private=<optimized out>, expected=14, futex_word=0xaafc60de87dc) at ../sysdeps/nptl/futex-internal.h:177
#2  __GI___pthread_cond_destroy (cond=0xaafc60de87b8) at ./nptl/pthread_cond_destroy.c:53
#3  0x0000aafc2f6190f8 in rstudio_boost::asio::detail::posix_event::~posix_event (this=0xaafc60de87b8, __in_chrg=<optimized out>)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/detail/posix_event.hpp:43
#4  0x0000aafc2f61b9a4 in rstudio_boost::asio::detail::conditionally_enabled_event::~conditionally_enabled_event (this=0xaafc60de87b0, __in_chrg=<optimized out>)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/detail/conditionally_enabled_event.hpp:44
#5  0x0000aafc2f61ed1c in rstudio_boost::asio::detail::scheduler::~scheduler (this=0xaafc60de8740, __in_chrg=<optimized out>)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/detail/impl/scheduler.ipp:152
#6  0x0000aafc2f61ed74 in rstudio_boost::asio::detail::scheduler::~scheduler (this=0xaafc60de8740, __in_chrg=<optimized out>)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/detail/impl/scheduler.ipp:152
#7  0x0000aafc2f61ab70 in rstudio_boost::asio::detail::service_registry::destroy (service=0xaafc60de8740) at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/detail/impl/service_registry.ipp:110
#8  0x0000aafc2f61aa70 in rstudio_boost::asio::detail::service_registry::destroy_services (this=0xaafc60de86f0)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/detail/impl/service_registry.ipp:54
#9  0x0000aafc2f61af84 in rstudio_boost::asio::execution_context::destroy (this=0xaafc30c17c10 <(anonymous namespace)::s_monitorIoService>)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/impl/execution_context.ipp:46
#10 0x0000aafc2f61af10 in rstudio_boost::asio::execution_context::~execution_context (this=0xaafc30c17c10 <(anonymous namespace)::s_monitorIoService>, __in_chrg=<optimized out>)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/impl/execution_context.ipp:35
#11 0x0000aafc2f61fb10 in rstudio_boost::asio::io_context::~io_context (this=0xaafc30c17c10 <(anonymous namespace)::s_monitorIoService>, __in_chrg=<optimized out>)
    at /opt/rstudio-tools/aarch64/boost/boost_1_83_0/include/boost/asio/impl/io_context.ipp:59
#12 0x0000f504e0e8cde8 in __run_exit_handlers (status=1, listp=0xf504e0feb730 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at ./stdlib/exit.c:113
#13 0x0000f504e0e8cf0c in __GI_exit (status=<optimized out>) at ./stdlib/exit.c:143
#14 0x0000aafc301429c4 in rstudio::core::system::ChildProcess::run (this=0xffffc946aa20) at /home/kevin/rstudio/src/cpp/core/system/PosixChildProcess.cpp:836
#15 0x0000aafc300cf0d8 in rstudio::core::system::SyncChildProcess::run (this=0xffffc946aa20,
    input="---\ntitle: \"Untitled\"\noutput: html_document\ndate: \"2024-08-28\"\n---\n\n```{r setup, include=FALSE}\nknitr::opts_chunk$set(echo = TRUE)\n```\n\n## R Markdown\n\nThis is an R Markdown document. Markdown is a sim"..., pResult=0xffffc946af88) at /home/kevin/rstudio/src/cpp/core/include/core/system/ChildProcess.hpp:144
#16 0x0000aafc300c90c4 in rstudio::core::system::runProgram (executable="/opt/rstudio-tools/aarch64/quarto/bin/tools/pandoc", args=std::vector of length 11, capacity 12 = {...},
    input="---\ntitle: \"Untitled\"\noutput: html_document\ndate: \"2024-08-28\"\n---\n\n```{r setup, include=FALSE}\nknitr::opts_chunk$set(echo = TRUE)\n```\n\n## R Markdown\n\nThis is an R Markdown document. Markdown is a sim"..., options=..., pResult=0xffffc946af88) at /home/kevin/rstudio/src/cpp/core/system/Process.cpp:55
#17 0x0000aafc2f8b2f34 in rstudio::session::module_context::runPandoc (pandocPath="/opt/rstudio-tools/aarch64/quarto/bin/tools/pandoc", args=std::vector of length 8, capacity 8 = {...},
    input="---\ntitle: \"Untitled\"\noutput: html_document\ndate: \"2024-08-28\"\n---\n\n```{r setup, include=FALSE}\nknitr::opts_chunk$set(echo = TRUE)\n```\n\n## R Markdown\n\nThis is an R Markdown document. Markdown is a sim"..., options=..., pResult=0xffffc946af88) at /home/kevin/rstudio/src/cpp/session/SessionPandoc.cpp:112
#18 0x0000aafc2fdd5564 in rstudio::session::(anonymous namespace)::indexSourceFile (
    contents="---\ntitle: \"Untitled\"\noutput: html_document\ndate: \"2024-08-28\"\n---\n\n```{r setup, include=FALSE}\nknitr::opts_chunk$set(echo = TRUE)\n```\n\n## R Markdown\n\nThis is an R Markdown document. Markdown is a sim"..., filename="zzz.Rmd") at /home/kevin/rstudio/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp:259
#19 0x0000aafc2fdd6908 in rstudio::session::(anonymous namespace)::resolvedXRefIndex (renderedIndexPath=..., srcPath=..., filename="zzz.Rmd")
    at /home/kevin/rstudio/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp:374
#20 0x0000aafc2fdd8394 in rstudio::session::modules::quarto::xrefs::(anonymous namespace)::xrefIndexForFile (filePath=..., indexJson=...)
    at /home/kevin/rstudio/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp:696
#21 0x0000aafc2fdd8a20 in rstudio::session::modules::quarto::xrefs::(anonymous namespace)::quartoXRefForId (request=..., pResponse=0xffffc946b848)
    at /home/kevin/rstudio/src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp:740
```

**Comment by kevinushey** [comment]
It looks like the call to `::fork()` (which is used to launch Pandoc) is failing. When we try to call `::exit()` in the forked sub-process, that call to `::exit()` hangs, waiting on the destructors of some Boost ASIO machinery. The parent session is then stuck waiting for a failed pandoc process that refuses to die.

The call to `::exit()` is here:

https://github.com/rstudio/rstudio/blob/a070094f53c40c71cf3cef51197837ac01877cff/src/cpp/core/system/PosixChildProcess.cpp#L828-L837

But rather than logging an error there, we should check the return status of our `::execve()` calls, and log an appropriate error.

A workaround is to use `::_exit()` instead of `::exit()` in fork, but I'm not certain what the ramifications of that are. I'll continue to poke around.

> 2024-08-29T00:42:49.356429Z [rsession-kevin] ERROR 2024-08-29T00:42:49.354169Z [rsession-kevin] ERROR system error 2 (No such file or directory) [exe: /opt/rstudio-tools/aarch64/quarto/bin/tools/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835; LOGGED FROM: rstudio::core::json::Array rstudio::session::{anonymous}::indexSourceFile(const string&, const string&) src/cpp/session/modules/quarto/SessionQuartoXRefs.cpp:274

**Comment by kevinushey** [comment]
Ah, and the executable path is incorrect. It should be:

> /opt/rstudio-tools/aarch64/quarto/bin/tools/aarch64/pandoc

but we're missing the architecture specifier.

**Comment by ronblum** [comment]
Verified in RStudio Desktop 2024.09.0+360 on Windows 11. Will check MacOS and Linux next, plus RStudio Server.

Tested with this reprex (stolen or borrowed from @gtritchie):

1. Create a new Quarto doc
2. Type `@abc` in the visual editor (e.g. on a new line after the Quarto heading)
3. Click on that text, nothing happens (e.g. no hang)
4. Save the file
5. Click on that `@abc` again, nothing happens (now that it's fixed)

Also tested with real citations from Zotero and DataCite:
1. Create a new Quarto doc
2. In visual mode: `Insert -> Citation` then insert a citations
3. Click on that text, the citation description appears
4. Save the file
5. Click on that `@abc` again, the citation description appears (now that it's fixed)

**Comment by ronblum** [comment]
Verified in RStudio Desktop 2024.09.0+360 in Ubuntu 22 and MacOS 15 beta.

**Comment by ronblum** [comment]
Verified in RStudio Server 2024.09.0+360 on Ubuntu 24 via Edge on Windows 11.

**Comment by raffaem** [comment]
> In my log I can find the following line:
> 
> > 2024-08-15T14:02:45.022581Z [rsession-basti] ERROR system error 2 (Datei oder Verzeichnis nicht gefunden) [exe: /usr/lib/rstudio/bin/pandoc/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835
> 
> When I checked the directory pandoc is not in
> 
> > /usr/lib/rstudio/bin/pandoc/
> 
> but in
> 
> > /usr/lib/rstudio/bin/pandoc/bin/tools/
> 
> After creating a soft link it works like a charm 🥳: `ln -s /usr/bin/pandoc /usr/lib/rstudio/bin/pandoc/pandoc`
> 
> I'm using Manjaro (arch based linux-distro), so maybe this info is helping out someone in a similar situation.

I'm gonna report it asa packaging bug

**Comment by raffaem** [comment]
> In my log I can find the following line:
> 
> > 2024-08-15T14:02:45.022581Z [rsession-basti] ERROR system error 2 (Datei oder Verzeichnis nicht gefunden) [exe: /usr/lib/rstudio/bin/pandoc/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835
> 
> When I checked the directory pandoc is not in
> 
> > /usr/lib/rstudio/bin/pandoc/
> 
> but in
> 
> > /usr/lib/rstudio/bin/pandoc/bin/tools/
> 
> After creating a soft link it works like a charm 🥳: `ln -s /usr/bin/pandoc /usr/lib/rstudio/bin/pandoc/pandoc`
> 
> I'm using Manjaro (arch based linux-distro), so maybe this info is helping out someone in a similar situation.



> In my log I can find the following line:
> 
> > 2024-08-15T14:02:45.022581Z [rsession-basti] ERROR system error 2 (Datei oder Verzeichnis nicht gefunden) [exe: /usr/lib/rstudio/bin/pandoc/pandoc]; OCCURRED AT rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:833; LOGGED FROM: rstudio::core::Error rstudio::core::system::ChildProcess::run() src/cpp/core/system/PosixChildProcess.cpp:835
> 
> When I checked the directory pandoc is not in
> 
> > /usr/lib/rstudio/bin/pandoc/
> 
> but in
> 
> > /usr/lib/rstudio/bin/pandoc/bin/tools/
> 
> After creating a soft link it works like a charm 🥳: `ln -s /usr/bin/pandoc /usr/lib/rstudio/bin/pandoc/pandoc`
> 
> I'm using Manjaro (arch based linux-distro), so maybe this info is helping out someone in a similar situation.

Actually the symlink I had to create was in a different path:

```
sudo ln -s /usr/bin/pandoc /opt/quarto/bin/tools/pandoc
```

I guess mine is for quarto document and yours for RMarkdown?

Or maybe manjaro install them in a different path?

**Comment by raffaem** [comment]
Actually Arch is just unpacking the official .deb file ([see PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=rstudio-desktop-bin)).

So I guess this is a bug of RStudio .deb packaging?

We are not providing pandoc in the package?

@ronblum shoud we reopen this?

**Comment by ronblum** [comment]
@raffaem Are you still getting the issue with the citations in visual mode or something else?

**Comment by raffaem** [comment]
> @raffaem Are you still getting the issue with the citations in visual mode or something else?

No because I fixed it myself in https://github.com/rstudio/rstudio/issues/10823#issuecomment-2336651984

**Comment by gtritchie** [comment]
@raffaem  The RStudio version on Arch is 2024.04.2-764 (per your "see PKGBUILD" link), so reopening the issue wouldn't make sense. The fix that addressed this issue is in a not-yet-released version of RStudio (2024.09.0).
